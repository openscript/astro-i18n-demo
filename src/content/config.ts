import { defineCollection, z, type ImageFunction } from 'astro:content';
import { C } from '../configuration';

const blogCollection = defineCollection({
  schema: ({ image }) => z.object({
    title: z.string(),
    publishedAt: z.date(),
    cover: z.object({
      src: image().refine((img) => img.width >= 800, {
        message: "Cover image must be at least 800 pixels wide!",
      }),
      alt: z.string().optional(),
    }).optional(),
  })
});
const createGalleryCollection = (image: ImageFunction) => z.object({
  title: z.string(),
  cover: image(),
  images: z.array(
    z.object({
      src: image().refine((img) => img.width >= 800, {
        message: "Image must be at least 800 pixels wide!",
      }),
      alt: z.string().optional(),
    })
  ),
}).optional();

const galleryCollection = defineCollection({
  type: 'data',
  schema: ({ image }) => z.object({...Object.keys(C.LOCALES).reduce<Record<string, ReturnType<typeof createGalleryCollection>>>((acc, locale) => ({...acc, [locale]: createGalleryCollection(image)}), {})}),
});
const pagesCollection = defineCollection({
  schema: z.object({
    path: z.string()
  })
});

export const collections = {
  'blog': blogCollection,
  'gallery': galleryCollection,
  'pages': pagesCollection
};
