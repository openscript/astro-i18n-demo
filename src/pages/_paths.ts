import type { GetStaticPaths } from 'astro';
import { localeSlugs, type Locale } from '../configuration';
import { getLocaleSlug, getMessage } from '../utils/i18n';
import { resolvePath } from '../utils/path';

export const rssXmlPaths = (async () => {
  const versions = [undefined, ...localeSlugs];
  return versions.map((l) => ({ params: { locale: l ? `-${l}` : undefined }, props: { locale: l } }));
}) satisfies GetStaticPaths;

export const indexPaths = (kind?: string) => {
  return (async () => {
    const translations = localeSlugs.reduce((acc, curr) => {
      const collectionSlug = kind ? getMessage(`slugs.${kind}`, curr) : undefined;
      const localeSlug = getLocaleSlug(curr);
      acc[curr] = resolvePath(localeSlug, collectionSlug);
      return acc;
    }, {} as Record<Locale, string>);
    return localeSlugs.map((l) => {
      const localeSlug = getLocaleSlug(l);
      const path = { params: { locale: localeSlug }, props: { locale: l, translations } };
      return kind ? { ...path, params: { ...path.params, [kind]: getMessage(`slugs.${kind}`, l) } } : path;
    });
  }) satisfies GetStaticPaths;
};



