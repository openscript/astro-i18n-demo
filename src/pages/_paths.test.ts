import { describe, expect, it, vi } from 'vitest';
import { entryPaths, indexPaths, rssXmlPaths } from './_paths';
import { getCollection } from 'astro:content';

vi.mock("../configuration", () => ({
  C: {
    LOCALES: { 'en': 'en-US', 'de': 'de-CH' },
    DEFAULT_LOCALE: 'en' as const,
    MESSAGES: {
      'en': {
        'language': 'English',
        'slugs.docs': 'docs',
        'slugs.data': 'data',
      },
      'de': {
        'language': 'Deutsch',
        'slugs.docs': 'doku',
        'slugs.data': 'daten',
      }
    }
  },
  localeSlugs: ['en', 'de']
}));

vi.mock("astro:content", () => ({
  getCollection: async (_: string) => {
    return [
      {
        id: "de/2020/09/11/test-article.md",
        slug: "de/2020/09/11/test-article",
        collection: "docs",
        data: {
          title: "Testartikel",
        }
      },
      {
        id: "en/2020/09/11/test-article.md",
        slug: "en/2020/09/11/test-article",
        collection: "docs",
        data: {
          title: "Test Article",
        }
      },
    ]
  }
}));

describe('paths', () => {
  it('generates rssXmlPaths', async () => {
    expect(await rssXmlPaths()).toMatchSnapshot();
  });
  it('generates indexPaths', async () => {
    expect(await indexPaths()()).toMatchSnapshot();
  });
  it('generates indexPaths with kind', async () => {
    expect(await indexPaths('docs')()).toMatchSnapshot();
  });
  //it('generates entryPaths', async () => {
  //  const entries = (await getCollection('docs' as any)) as { slug: string }[];
  //  expect(await entryPaths('blog')()).toMatchSnapshot();
  //});
})
