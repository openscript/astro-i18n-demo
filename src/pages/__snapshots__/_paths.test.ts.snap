// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`paths > generates entryPaths 1`] = `
[
  {
    "params": {
      "locale": "de",
      "pages": "2020/09/11/testartikel",
    },
    "props": {
      "collection": "pages",
      "data": {
        "title": "Testartikel",
      },
      "id": "de/2020/09/11/test-article.md",
      "locale": "de",
      "slug": "de/2020/09/11/test-article",
      "translations": {
        "de": "/de/2020/09/11/testartikel",
        "en": "/2020/09/11/test-article",
      },
    },
  },
  {
    "params": {
      "locale": undefined,
      "pages": "2020/09/11/test-article",
    },
    "props": {
      "collection": "pages",
      "data": {
        "title": "Test Article",
      },
      "id": "en/2020/09/11/test-article.md",
      "locale": "en",
      "slug": "en/2020/09/11/test-article",
      "translations": {
        "de": "/de/2020/09/11/testartikel",
        "en": "/2020/09/11/test-article",
      },
    },
  },
]
`;

exports[`paths > generates entryPaths 2`] = `
[
  {
    "params": {
      "docs": "doku",
      "locale": "de",
      "slug": "2020/09/11/testartikel",
    },
    "props": {
      "collection": "docs",
      "data": {
        "title": "Testartikel",
      },
      "id": "de/2020/09/11/test-article.md",
      "locale": "de",
      "slug": "de/2020/09/11/test-article",
      "translations": {
        "de": "/de/doku/2020/09/11/testartikel",
        "en": "/docs/2020/09/11/test-article",
      },
    },
  },
  {
    "params": {
      "docs": "docs",
      "locale": undefined,
      "slug": "2020/09/11/test-article",
    },
    "props": {
      "collection": "docs",
      "data": {
        "title": "Test Article",
      },
      "id": "en/2020/09/11/test-article.md",
      "locale": "en",
      "slug": "en/2020/09/11/test-article",
      "translations": {
        "de": "/de/doku/2020/09/11/testartikel",
        "en": "/docs/2020/09/11/test-article",
      },
    },
  },
]
`;

exports[`paths > generates galleryCategoryPaths 1`] = `
[
  {
    "params": {
      "category": "food",
      "gallery": "gallery",
      "locale": undefined,
    },
    "props": {
      "locale": "en",
      "page": {
        "cover": "./food1.jpg",
        "images": [
          {
            "src": "./food1.jpg",
          },
          {
            "src": "./food2.jpg",
          },
          {
            "src": "./food3.jpg",
          },
        ],
        "title": {
          "de": "Essen",
          "en": "Food",
        },
      },
      "translations": {
        "de": "/de/galerie/essen",
        "en": "/gallery/food",
      },
    },
  },
  {
    "params": {
      "category": "essen",
      "gallery": "galerie",
      "locale": "de",
    },
    "props": {
      "locale": "de",
      "page": {
        "cover": "./food1.jpg",
        "images": [
          {
            "src": "./food1.jpg",
          },
          {
            "src": "./food2.jpg",
          },
          {
            "src": "./food3.jpg",
          },
        ],
        "title": {
          "de": "Essen",
          "en": "Food",
        },
      },
      "translations": {
        "de": "/de/galerie/essen",
        "en": "/gallery/food",
      },
    },
  },
  {
    "params": {
      "category": "sea",
      "gallery": "gallery",
      "locale": undefined,
    },
    "props": {
      "locale": "en",
      "page": {
        "cover": "./sea1.jpg",
        "images": [
          {
            "src": "./sea1.jpg",
          },
          {
            "src": "./sea2.jpg",
          },
          {
            "src": "./sea3.jpg",
          },
          {
            "src": "./sea4.jpg",
          },
        ],
        "title": {
          "de": "Meer",
          "en": "Sea",
        },
      },
      "translations": {
        "de": "/de/galerie/meer",
        "en": "/gallery/sea",
      },
    },
  },
  {
    "params": {
      "category": "meer",
      "gallery": "galerie",
      "locale": "de",
    },
    "props": {
      "locale": "de",
      "page": {
        "cover": "./sea1.jpg",
        "images": [
          {
            "src": "./sea1.jpg",
          },
          {
            "src": "./sea2.jpg",
          },
          {
            "src": "./sea3.jpg",
          },
          {
            "src": "./sea4.jpg",
          },
        ],
        "title": {
          "de": "Meer",
          "en": "Sea",
        },
      },
      "translations": {
        "de": "/de/galerie/meer",
        "en": "/gallery/sea",
      },
    },
  },
  {
    "params": {
      "category": "space",
      "gallery": "gallery",
      "locale": undefined,
    },
    "props": {
      "locale": "en",
      "page": {
        "cover": "./space1.jpg",
        "images": [
          {
            "src": "./space1.jpg",
          },
          {
            "src": "./space2.jpg",
          },
          {
            "src": "./space3.jpg",
          },
          {
            "src": "./space4.jpg",
          },
          {
            "src": "./space5.jpg",
          },
        ],
        "title": {
          "de": "Weltraum",
          "en": "Space",
        },
      },
      "translations": {
        "de": "/de/galerie/weltraum",
        "en": "/gallery/space",
      },
    },
  },
  {
    "params": {
      "category": "weltraum",
      "gallery": "galerie",
      "locale": "de",
    },
    "props": {
      "locale": "de",
      "page": {
        "cover": "./space1.jpg",
        "images": [
          {
            "src": "./space1.jpg",
          },
          {
            "src": "./space2.jpg",
          },
          {
            "src": "./space3.jpg",
          },
          {
            "src": "./space4.jpg",
          },
          {
            "src": "./space5.jpg",
          },
        ],
        "title": {
          "de": "Weltraum",
          "en": "Space",
        },
      },
      "translations": {
        "de": "/de/galerie/weltraum",
        "en": "/gallery/space",
      },
    },
  },
]
`;

exports[`paths > generates indexPaths 1`] = `
[
  {
    "params": {
      "locale": undefined,
    },
    "props": {
      "locale": "en",
      "translations": {
        "de": "/de",
        "en": "/",
      },
    },
  },
  {
    "params": {
      "locale": "de",
    },
    "props": {
      "locale": "de",
      "translations": {
        "de": "/de",
        "en": "/",
      },
    },
  },
]
`;

exports[`paths > generates indexPaths with kind 1`] = `
[
  {
    "params": {
      "docs": "docs",
      "locale": undefined,
    },
    "props": {
      "locale": "en",
      "translations": {
        "de": "/de/doku",
        "en": "/docs",
      },
    },
  },
  {
    "params": {
      "docs": "doku",
      "locale": "de",
    },
    "props": {
      "locale": "de",
      "translations": {
        "de": "/de/doku",
        "en": "/docs",
      },
    },
  },
]
`;

exports[`paths > generates rssXmlPaths 1`] = `
[
  {
    "params": {
      "locale": undefined,
    },
    "props": {
      "locale": undefined,
    },
  },
  {
    "params": {
      "locale": "-en",
    },
    "props": {
      "locale": "en",
    },
  },
  {
    "params": {
      "locale": "-de",
    },
    "props": {
      "locale": "de",
    },
  },
]
`;
